pkg/test/parameter_test.go

```diff
@@ -7,46 +7,54 @@ package test
 
 import (
 	"errors"
+	"fmt"
 	"strings"
 	"testing"
 
 	"github.com/facebookincubator/contest/pkg/target"
+	"github.com/facebookincubator/contest/pkg/types"
 	"github.com/stretchr/testify/require"
 )
 
 func TestParameterExpand(t *testing.T) {
 	validExprs := [][4]string{
-		// expression, target name, target ID, expected result
-		[4]string{"{{ ToLower .Name }}", "www.slackware.IT", "1234", "www.slackware.it"},
-		[4]string{"{{ .Name }}", "www.slackware.IT", "2345", "www.slackware.IT"},
-		[4]string{"{{ .Name }}", "www.slackware.it", "3456", "www.slackware.it"},
-		[4]string{"name={{ .Name }}, id={{ .ID }}", "blah", "12345", "name=blah, id=12345"},
+		// expression, target FQDN, target ID, expected result
+		[4]string{"{{ ToLower .FQDN }}", "www.slackware.IT", "1234", "www.slackware.it"},
+		[4]string{"{{ .FQDN }}", "www.slackware.IT", "2345", "www.slackware.IT"},
+		[4]string{"{{ .FQDN }}", "www.slackware.it", "3456", "www.slackware.it"},
+		[4]string{"fqdn={{ .FQDN }}, id={{ .ID }}", "blah", "12345", "fqdn=blah, id=12345"},
 	}
 	for _, x := range validExprs {
 		p := NewParam(x[0])
-		res, err := p.Expand(&target.Target{Name: x[1], ID: x[2]})
+		res, err := p.Expand(&target.Target{FQDN: x[1], ID: types.TargetID(x[2])})
 		require.NoError(t, err, x[0])
 		require.Equal(t, x[3], res, x[0])
 	}
 }
 
 func TestParameterExpandUserFunctions(t *testing.T) {
-	title := func(a ...string) (string, error) {
+	customFunc := func(a ...interface{}) (string, error) {
 		if len(a) == 0 {
 			return "", errors.New("no params")
 		}
-		return strings.Title(a[0]), nil
+		switch v := a[0].(type) {
+		case string:
+			return strings.Title(v), nil
+		case types.TargetID:
+			return strings.Title(string(v)), nil
+		}
+		return "", fmt.Errorf("unexpected input type")
 	}
-	err := RegisterFunction("title", title)
+	err := RegisterFunction("CustomFunc", customFunc)
 	require.NoError(t, err)
 	validExprs := [][4]string{
-		// expression, target name, target ID, expected result
-		[4]string{"{{ Title .Name }}", "slackware.it", "1234", "Slackware.It"},
-		[4]string{"{{ Title .ID }}", "slackware.it", "a1234a", "A1234a"},
+		// expression, target FQDN, target ID, expected result
+		[4]string{"{{ CustomFunc .FQDN }}", "slackware.it", "1234", "Slackware.It"},
+		[4]string{"{{ CustomFunc .ID }}", "slackware.it", "a1234a", "A1234a"},
```

Comment:
We can't use a `func(a ...string) (string, error)` anymore if we want to template `func .ID`. The template parsing/execution doesn't understand and breaks:
```
--- FAIL: TestParameterExpandUserFunctions (0.00s)
    parameter_test.go:58:
        	Error Trace:	parameter_test.go:53
        	Error:      	Received unexpected error:
        	            	template: :1:9: executing "" at <.ID>: wrong type for value; expected string; got types.TargetID
        	Test:       	TestParameterExpandUserFunctions
        	Messages:   	{{ Title .ID }}
FAIL
```
The workaround is accepting `interface` and doing a type switch, but this is ugly and means we can no longer use simple things like `strings.ToUpper`

Reply:
I am for reverting from `types.TargetID` to `string`, it gives us more trouble than benefits